### PR TITLE
Indavideo token

### PR DESCRIPTION
**Pull request by dnet**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

It seems that Indavideo has added an extra token to their JSON metadata, and if the token is not used as a URL parameter when requesting the actual video file, the request is denied with a 403 Forbidden response. This pull request extracts the token and appends it to the video URL transparently, using the standard Python 2/3 URL parsing tools to be as future-proof as possible.

### CODE REVIEW

File: youtube_dl/extractor/indavideo.py [hunk 1]
```diff
@@ -1,6 +1,14 @@
 # coding: utf-8
 from __future__ import unicode_literals
 
+try:
+    # Python 3
+    from urllib.parse import urlsplit, urlunsplit, urlencode, parse_qsl
+except ImportError:
+    # Python 2
+    from urlparse import urlsplit, urlunsplit, parse_qsl
+    from urllib import urlencode
```
Comment (dstftw) [inline youtube_dl/extractor/indavideo.py]: `compat_*` functions.

File: youtube_dl/extractor/indavideo.py [hunk 2]
```diff
@@ -90,6 +95,20 @@ def _real_extract(self, url):
             'formats': formats,
         }
 
+    def video_url_to_format(self, video_url, filesh):
+        height = int_or_none(self._search_regex(
+            r'\.(\d{3,4})\.mp4(?:\?|$)', video_url, 'height', default=None))
+        if height is not None and filesh is not None:
+            token = filesh.get(str(height))
```
Comment (dstftw) [inline youtube_dl/extractor/indavideo.py]: `compat_str`.

File: youtube_dl/extractor/indavideo.py [hunk 3]
```diff
@@ -90,6 +95,20 @@ def _real_extract(self, url):
...
+            if token is not None:
+                us = urlsplit(video_url)
+                query = urlencode(parse_qsl(us.query) + [('token', token)])
+                video_url = urlunsplit((us.scheme, us.netloc, us.path, query, us.fragment))
```
Comment (dstftw) [inline youtube_dl/extractor/indavideo.py]: `update_url_query`.

File: youtube_dl/extractor/indavideo.py [hunk 4]
```diff
@@ -90,6 +95,20 @@ def _real_extract(self, url):
             'formats': formats,
         }
 
+    def video_url_to_format(self, video_url, filesh):
+        height = int_or_none(self._search_regex(
+            r'\.(\d{3,4})\.mp4(?:\?|$)', video_url, 'height', default=None))
+        if height is not None and filesh is not None:
```
Comment (dstftw) [inline youtube_dl/extractor/indavideo.py]: `if height and filesh:`

**Comment by dnet** [comment]
@dstftw wrote

> Does not work:

You invoked a URL with Indavideo directly, I submitted patches regarding embed URLs. So my PR didn't break existing working code, but rather fixed another part. Here's an example that doesn't work with the current `HEAD` but it works with my patch applied: https://embed.indavideo.hu/player/video/c8f72ea6ef

File: youtube_dl/extractor/indavideo.py (hunk 4 above)
Comment (dnet) [reply youtube_dl/extractor/indavideo.py]: ACK acb14703b

File: youtube_dl/extractor/indavideo.py (hunk 2 above)
Comment (dnet) [reply youtube_dl/extractor/indavideo.py]: ACK 425f79940

File: youtube_dl/extractor/indavideo.py (hunk 1 above)
Comment (dnet) [reply youtube_dl/extractor/indavideo.py]: not necessary thanks to your helpful suggestion of `update_url_query` below

File: youtube_dl/extractor/indavideo.py (hunk 3 above)
Comment (dnet) [reply youtube_dl/extractor/indavideo.py]: ACK f1a6deb23

**Comment by dnet** [comment]
Ping?

**Comment by dnet** [comment]
Is there a problem with my PR?

**Pull request merged**